File: tools/legacy-reference/lib/intelligence/review-metrics.ts
// FROZEN COPY from Mitosia/mitosia-legacy `lib/intelligence/review-metrics.ts`
// at commit b642b774b48ad45910acb168d8dad86796da7e64, taken 2026-09-07.
// Unmodified; it imports nothing. Here for the record beside the scorers: the
// Python port under apps/pipeline/src/temnia_pipeline/evals/ is asserted against
// the committed parity snapshots, not against this file. See ./README.md.

// The Gate M1 readout's pure half: acceptance rate and boundary-adjustment
// magnitude over reviewed clip rows (S6 §7). Extracted from
// scripts/moment-metrics.ts so the Python pipeline's eval runner can port
// these exact semantics and hold parity against them (pipeline A1); the
// script keeps the SQL and rendering loop, this module owns the math and
// the report text.

export interface ReviewedClipRow {
  adjustedEndMs: number | null;
  adjustedStartMs: number | null;
  endMs: number;
  startMs: number;
  status: string;
}

export interface ReviewMetrics {
  accepted: number;
  deltasIn: number[];
  deltasOut: number[];
  proposed: number;
  rejected: number;
  shortlisted: number;
  total: number;
}

export function emptyReviewMetrics(): ReviewMetrics {
  return {
    accepted: 0,
    deltasIn: [],
    deltasOut: [],
    proposed: 0,
    rejected: 0,
    shortlisted: 0,
    total: 0,
  };
}

export function accumulateReviewMetrics(
  metrics: ReviewMetrics,
  row: ReviewedClipRow
): void {
  metrics.total += 1;
  if (row.status === "accepted") {
    metrics.accepted += 1;
    metrics.deltasIn.push(
      Math.abs((row.adjustedStartMs ?? row.startMs) - row.startMs)
    );
    metrics.deltasOut.push(
      Math.abs((row.adjustedEndMs ?? row.endMs) - row.endMs)
    );
  } else if (row.status === "rejected") {
    metrics.rejected += 1;
  } else if (row.status === "shortlisted") {
    metrics.shortlisted += 1;
  } else {
    metrics.proposed += 1;
  }
}

export function median(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  const upper = sorted[middle] ?? 0;
  if (sorted.length % 2 === 1) {
    return upper;
  }
  return ((sorted[middle - 1] ?? 0) + upper) / 2;
}

export function mean(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  return values.reduce((total, value) => total + value, 0) / values.length;
}

function seconds(ms: number): string {
  return `${(ms / 1000).toFixed(1)}s`;
}

export function reviewMetricsLines(
  label: string,
  metrics: ReviewMetrics
): string[] {
  const decided = metrics.accepted + metrics.rejected;
  const acceptance =
    decided === 0
      ? "no decisions yet"
      : `${metrics.accepted}/${decided} (${((metrics.accepted / decided) * 100).toFixed(0)}%)`;
  const lines = [
    `${label}`,
    `  candidates: ${metrics.total} — ${metrics.accepted} accepted · ${metrics.shortlisted} shortlisted · ${metrics.rejected} rejected · ${metrics.proposed} undecided`,
    `  acceptance rate: ${acceptance}`,
  ];
  if (metrics.accepted > 0) {
    lines.push(
      `  boundary Δ (accepted) — in: mean ${seconds(mean(metrics.deltasIn))}, median ${seconds(median(metrics.deltasIn))} · out: mean ${seconds(mean(metrics.deltasOut))}, median ${seconds(median(metrics.deltasOut))}`
    );
  }
  return lines;
}

File: apps/pipeline/src/temnia_pipeline/evals/review_metrics.py
"""The Gate M1 readout's pure half, ported from ``lib/intelligence/review-metrics.ts``.

Acceptance rate and boundary-adjustment magnitude over reviewed clip rows;
report lines match the TS text byte-for-byte (the parity snapshots assert
it). The SQL feeding these rows arrives with the A2 database layer.
"""

from collections.abc import Sequence

from pydantic import Field

from temnia_pipeline.evals.wire import WireModel
from temnia_pipeline.jsnum import to_fixed


class ReviewedClipRow(WireModel):
    adjusted_end_ms: int | None
    adjusted_start_ms: int | None
    end_ms: int
    start_ms: int
    status: str


class ReviewMetrics(WireModel):
    accepted: int = 0
    deltas_in: list[int] = Field(default_factory=list[int])
    deltas_out: list[int] = Field(default_factory=list[int])
    proposed: int = 0
    rejected: int = 0
    shortlisted: int = 0
    total: int = 0


def accumulate_review_metrics(metrics: ReviewMetrics, row: ReviewedClipRow) -> None:
    metrics.total += 1
    if row.status == "accepted":
        metrics.accepted += 1
        adjusted_start = (
            row.adjusted_start_ms if row.adjusted_start_ms is not None else row.start_ms
        )
        adjusted_end = row.adjusted_end_ms if row.adjusted_end_ms is not None else row.end_ms
        metrics.deltas_in.append(abs(adjusted_start - row.start_ms))
        metrics.deltas_out.append(abs(adjusted_end - row.end_ms))
    elif row.status == "rejected":
        metrics.rejected += 1
    elif row.status == "shortlisted":
        metrics.shortlisted += 1
    else:
        metrics.proposed += 1


def median(values: Sequence[float]) -> float:
    if len(values) == 0:
        return 0
    ordered = sorted(values)
    middle = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[middle]
    return (ordered[middle - 1] + ordered[middle]) / 2


def mean(values: Sequence[float]) -> float:
    if len(values) == 0:
        return 0
    total = 0.0
    for value in values:
        total += value
    return total / len(values)


def _seconds(ms: float) -> str:
    return f"{to_fixed(ms / 1000, 1)}s"


def review_metrics_lines(label: str, metrics: ReviewMetrics) -> list[str]:
    decided = metrics.accepted + metrics.rejected
    acceptance = (
        "no decisions yet"
        if decided == 0
        else (f"{metrics.accepted}/{decided} ({to_fixed((metrics.accepted / decided) * 100, 0)}%)")
    )
    lines = [
        f"{label}",
        (
            f"  candidates: {metrics.total} — {metrics.accepted} accepted"
            f" · {metrics.shortlisted} shortlisted · {metrics.rejected} rejected"
            f" · {metrics.proposed} undecided"
        ),
        f"  acceptance rate: {acceptance}",
    ]
    if metrics.accepted > 0:
        lines.append(
            f"  boundary Δ (accepted) — in: mean {_seconds(mean(metrics.deltas_in))},"
            f" median {_seconds(median(metrics.deltas_in))}"
            f" · out: mean {_seconds(mean(metrics.deltas_out))},"
            f" median {_seconds(median(metrics.deltas_out))}"
        )
    return lines
